File: kotlin/src/main/kotlin/Day01.kt
fun main() {
  val input = util.readInput("Day01.txt")

  fun part1(input: List<String>): Int =
      input.sumOf { cal ->
        cal.first { it.isDigit() }.digitToInt() * 10 + cal.last { it.isDigit() }.digitToInt()
      }

  fun part2(input: List<String>): Int {
    val reprToDigit =
        mapOf(
            "one" to 1,
            "two" to 2,
            "three" to 3,
            "four" to 4,
            "five" to 5,
            "six" to 6,
            "seven" to 7,
            "eight" to 8,
            "nine" to 9
        ) + (1..9).map { it.toString() to it }

    fun numberFromLine(line: String): Int {
      val capturingRegex = ("(?=(" + reprToDigit.keys.joinToString("|") + "))").toRegex()
      val matches = capturingRegex.findAll(line).map { it.groupValues[1] }.toList()
      val firstDigit = reprToDigit[matches.first()]!!
      val lastDigit = reprToDigit[matches.last()]!!

      return firstDigit * 10 + lastDigit
    }

    return input.sumOf { numberFromLine(it) }
  }

  part1(input).dbg()
  part2(input).dbg()
}

File: scala/src/Day01.scala
import fs2.io.file.Files
import cats.effect.IO
import fs2.io.file.Path
import fs2.Stream
import cats.effect.IOApp

import utils.*

object Day01 extends IOApp.Simple:
  private def part1(input: Stream[IO, String]): Stream[IO, Int] =
    input
      .map: cal =>
        val digits = cal.filter(_.isDigit)
        s"${digits.head}${digits.last}".toInt
      .foldMonoid
      .head

  private def part2(input: Stream[IO, String]): Stream[IO, Int] =
    val digitRepr = Map(
      "one" -> 1,
      "two" -> 2,
      "three" -> 3,
      "four" -> 4,
      "five" -> 5,
      "six" -> 6,
      "seven" -> 7,
      "eight" -> 8,
      "nine" -> 9
    ) ++ (1 to 9).map(i => i.toString() -> i)

    def getDigits(line: String): Int =
      val patt = "one|two|three|four|five|six|seven|eight|nine"
      val startPatt = (patt ++ "|\\d").r
      val endPatt = (patt.reverse ++ "|\\d").r
      val startDigit =
        startPatt
          .findFirstMatchIn(line)
          .fold(0)(m => digitRepr(m.matched))
      val lastDigit =
        endPatt
          .findFirstMatchIn(line.reverse)
          .fold(0)(m => digitRepr(m.matched.reverse))

      startDigit * 10 + lastDigit

    input
      .map(getDigits)
      .foldMonoid
      .head

  def run: IO[Unit] =
    val input = readInput("Day01")

    part1(input)
      .zip(part2(input))
      .evalTap(IO.println)
      .compile
      .drain
